internal/cmd/local/k8s/client.go

```diff
@@ -69,7 +76,78 @@ var _ Client = (*DefaultK8sClient)(nil)
 
 // DefaultK8sClient converts the official kubernetes client to our more manageable (and testable) interface
 type DefaultK8sClient struct {
-	ClientSet *kubernetes.Clientset
+	ClientSet kubernetes.Interface
+}
+
+func (d *DefaultK8sClient) DeploymentRestart(ctx context.Context, namespace, name string) error {
+	return d.deploymentRestart(ctx, namespace, name, time.Now(), 5*time.Minute)
+}
+
+// internal function so the restartedAt value can be specified for testing purposes
+func (d *DefaultK8sClient) deploymentRestart(ctx context.Context, namespace, name string, restartedAt time.Time, block time.Duration) error {
+	restartedAtName := "kubectl.kubernetes.io/restartedAt"
+	restartedAtValue := restartedAt.Format(time.RFC3339)
+
+	// similar to how kubectl rollout restart works, patch in a restartedAt annotation.
+	rawPatch := map[string]any{
+		"spec": map[string]any{
+			"template": map[string]any{
+				"metadata": map[string]any{
+					"annotations": map[string]string{
+						restartedAtName: restartedAtValue,
+					},
+				},
+			},
+		},
+	}
+
+	jsonData, err := json.Marshal(rawPatch)
+	if err != nil {
+		return fmt.Errorf("unable to marshal raw patch: %w", err)
+	}
+
+	deployment, err := d.ClientSet.AppsV1().Deployments(namespace).Patch(ctx, name, types.StrategicMergePatchType, jsonData, metav1.PatchOptions{})
+	if err != nil {
+		return fmt.Errorf("unable to fetch deployment %s: %w", name, err)
+	}
+
+	label := metav1.FormatLabelSelector(deployment.Spec.Selector)
+
+	deploymentPods := func(ctx context.Context) (bool, error) {
+		pods, err := d.ClientSet.CoreV1().Pods(namespace).List(ctx, metav1.ListOptions{
+			LabelSelector: label,
+		})
+		if err != nil {
+			return false, fmt.Errorf("unable to list pods for deployment %s: %w", name, err)
+		}
+
+		for _, pod := range pods.Items {
+			// if any pods are not running or are missing the restartedAt annotation
+			// then the restart isn't complete
+			if pod.Status.Phase != corev1.PodRunning || pod.ObjectMeta.Annotations[restartedAtName] != restartedAtValue {
+				return false, nil
+			}
+
+			// even though a pod is running, doesn't mean it is ready
+			for _, condition := range pod.Status.Conditions {
+				if condition.Type == corev1.PodReady && condition.Status != corev1.ConditionTrue {
+					return false, nil
+				}
+			}
+		}
+
+		// if we're here, then all the pods are running with the correct restartedAt annotation,
+		// and they're in a ready state
+		return true, nil
+	}
+
+	// check every 10 seconds for up to 5 minutes to see if the pods have been restarted successfully
+	err = wait.PollUntilContextTimeout(ctx, 5*time.Second, block, true, deploymentPods)
```

Comment:
Comment does not match the invocation (interval is 5 not 10). Also (nit): there may be a clearer than for this arg than `block`